tests/test_project.py

```diff
@@ -871,6 +871,106 @@ def test_update_submodules_strategy(repos_tmpdir):
+    # Regression test for
+    # https://github.com/zephyrproject-rtos/west/issues/545.
+    #
+    # We need to make sure that submodules with relative paths are
+    # initialized correctly even when ther is no "origin" remote.
+    #
+    # Background: unless the config variable git.$SUBMODULENAME.url is
+    # set, git relies on the "default remote" for figuring out how to
+    # turn that relative path into something useful. The "default
+    # remote" ends up being "origin" in the situation where we are
+    # initializing a submodule from a superproject which is a west
+    # project. However, west neither makes nor expects any guarantees
+    # about any particular git remotes being present in a project, so
+    # that only works if we set the config variable before
+    # initializing the module.
+
+    # The following paths begin with 'project' if they are west projects,
+    # and 'submodule' if they are direct or transitive submodules of a
+    # west project. Numeric suffixes reflect the tree structure of the
+    # workspace:
+    #
+    # workspace
+    # ├── manifest
+    # ├── project-1
+    # │   └── submodule-1-1
+    # └── project-2
+    #     ├── submodule-2-1
+    #     └── submodule-2-2
+    #         └── submodule-2-2-1
+    remotes = tmpdir / 'remotes'  # parent directory for 'remote' repos
+    remote_repositories = [
+        remotes / path for path in
+        ['manifest',
+         'project-1',
+         'submodule-1-1',
+         'project-2',
+         'submodule-2-1',
+         'submodule-2-2',
+         'submodule-2-2-1',
+         ]
+    ]
+    for remote_repo in remote_repositories:
+        create_repo(remote_repo)
+
+    add_commit(remotes / 'manifest', 'manifest west.yml',
+               files={'west.yml': f'''
+manifest:
+  remotes:
+    - name: not-origin
+      url-base: {remotes}
+  defaults:
+    remote: not-origin
+  projects:
+    - name: project-1
+      submodules: true
+    - name: project-2
+      submodules:
+        - path: submodule-2-1
+        - name: sub-2-2
+          path: submodule-2-2
+'''})
+
+    def add_submodule(superproject, submodule_name,
+                      submodule_url, submodule_path):
+        subprocess.check_call([GIT, '-C', os.fspath(superproject),
+                               'submodule', 'add',
+                               '--name', submodule_name,
+                               submodule_url, submodule_path])
+        add_commit(superproject, f'add submodule {submodule_name}')
+
+    add_submodule(remotes / 'project-1', 'sub-1-1',
+                  '../submodule-1-1', 'submodule-1-1')
+    add_submodule(remotes / 'project-2', 'sub-2-1',
+                  '../submodule-2-1', 'submodule-2-1')
+    add_submodule(remotes / 'project-2', 'sub-2-2',
+                  '../submodule-2-2', 'submodule-2-2')
+    add_submodule(remotes / 'project-2' / 'submodule-2-2', 'sub-2-2-1',
+                  '../submodule-2-2-1', 'submodule-2-2-1')
```

Comment:
I think you need to add these from the bottom up, not top down otherwise the commits at the top will be lagging behind (git status should show that)

Reply:
I don't follow, sorry. How can I add a sub-submodule for a submodule that isn't added yet?

Reply:
OK so let's say we want A->B->C from parent to child. You certainly don't need A to add C to B (B->C). Plus it's faster to link B->C first because when you add B to A later then C comes with it "for free" without having to go back and perform another commit.

If you do A->B first, then after doing B->C then you need to go back to A to update B to the newer and "bigger" B that includes C.

BTW this test code does not seem to have any commit yet?

Reply:
> You certainly don't need A to add C to B (B->C).

I must be missing something. I think I definitely do need A to add C to B.

I need somewhere to add C. That place is B. I cannot create B without creating A, since B lives inside A.

What am I missing? Can you phrase this suggestion in the form of a diff?

Reply:
I'm not familiar with the repo initialization routines and I'm not sure I have the best solution, but the initialization problem I noticed when reading the code seems confirmed when trying to run it:
```
tox -- -k submodules_rel --runxfail
cd /tmp/blah/pytest-6/test_update_submodules_relpath0/remotes/

tree project-2

project-2/
├── submodule-2-1
└── submodule-2-2
    └── submodule-2-2-1

git -C project-2/ status

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   submodule-2-2 (new commits)

git clone --recursive project-2 incomplete

tree incomplete/
incomplete/
├── submodule-2-1
└── submodule-2-2
```
Also,  the remotes should all be at the same level. In a real-world example they would all be bare repos at the same level. Relative paths are complicated enough, better not add small divergences. It may be possible to simply move the remotes after initialization. As long as they're only used as remotes, only their `.git/` will be used so it should hopefully not matter whether the checkout is in some inconsistent state.

Reply:
> I cannot create B without creating A, since B lives inside A

I think you're confusing or conflating initial repo creation with adding a submodule.

```
$ tree
├── A
├── B
└── C

$ git -C B submodule add ../C
Cloning into 'submtest/test/B/C'...

$ git -C B commit -m 'add ../C to B'
[master 616d32033d4a] add ../C to B
 2 files changed, 4 insertions(+)
 create mode 100644 .gitmodules
 create mode 160000 C

$ git -C A submodule add ../B
Cloning into 'submtest/test/A/B'...
done.

$ tree A/
A/
└── B
    └── C

```